File: cpp/basic/image_features.cpp
#include <basic/image_features.hpp>
#include <cmath>

namespace ip101 {

using namespace cv;
using namespace std;

namespace {
// Internal constants
constexpr int CACHE_LINE = 64;    // CPU cache line size (bytes)
constexpr int SIMD_WIDTH = 32;    // AVX2 SIMD vector width (bytes)
constexpr int BLOCK_SIZE = 16;    // Block processing size
constexpr float PI_F = 3.14159265358979323846f;  // PI as float
constexpr double PI = 3.14159265358979323846;    // PI as double

// Memory alignment helper function
inline uchar* alignPtr(uchar* ptr, size_t align = CACHE_LINE) {
    return (uchar*)(((size_t)ptr + align - 1) & ~(align - 1));
}

// Calculate gradient
void compute_gradient(const Mat& src, Mat& magnitude, Mat& angle) {
    Mat grad_x, grad_y;
    Sobel(src, grad_x, CV_32F, 1, 0, 3);
    Sobel(src, grad_y, CV_32F, 0, 1, 3);

    magnitude.create(src.size(), CV_32F);
    angle.create(src.size(), CV_32F);

    #pragma omp parallel for
    for (int y = 0; y < src.rows; y++) {
        for (int x = 0; x < src.cols; x++) {
            float gx = grad_x.at<float>(y, x);
            float gy = grad_y.at<float>(y, x);
            magnitude.at<float>(y, x) = sqrt(gx * gx + gy * gy);
            angle.at<float>(y, x) = atan2(gy, gx);
        }
    }
}

} // anonymous namespace

void hog_features(const Mat& src,
                 vector<float>& features,
                 int cell_size,
                 int block_size,
                 int bin_num) {
    CV_Assert(!src.empty());

    // Convert to grayscale
    Mat gray;
    if (src.channels() == 3) {
        cvtColor(src, gray, COLOR_BGR2GRAY);
    } else {
        gray = src.clone();
    }

    // Calculate gradient
    Mat magnitude, angle;
    compute_gradient(gray, magnitude, angle);

    // Calculate cell histograms
    int cell_rows = gray.rows / cell_size;
    int cell_cols = gray.cols / cell_size;
    vector<vector<vector<float>>> cell_hists(cell_rows,
        vector<vector<float>>(cell_cols, vector<float>(bin_num, 0)));

    #pragma omp parallel for
    for (int y = 0; y < gray.rows - cell_size; y += cell_size) {
        for (int x = 0; x < gray.cols - cell_size; x += cell_size) {
            vector<float> hist(bin_num, 0);

            // Calculate gradient histogram within cell
            for (int cy = 0; cy < cell_size; cy++) {
                for (int cx = 0; cx < cell_size; cx++) {
                    float mag = magnitude.at<float>(y + cy, x + cx);
                    float ang = angle.at<float>(y + cy, x + cx);
                    if (ang < 0) ang += static_cast<float>(PI);

                    float bin_size = static_cast<float>(PI) / static_cast<float>(bin_num);
                    int bin = static_cast<int>(ang / bin_size);
                    if (bin >= bin_num) bin = bin_num - 1;

                    hist[bin] += mag;
                }
            }

            cell_hists[y/cell_size][x/cell_size] = hist;
        }
    }

    // Calculate block features
    features.clear();
    for (int y = 0; y <= cell_rows - block_size; y++) {
        for (int x = 0; x <= cell_cols - block_size; x++) {
            vector<float> block_feat;
            float norm = 0.0f;

            // Collect all cell histograms within the block
            for (int by = 0; by < block_size; by++) {
                for (int bx = 0; bx < block_size; bx++) {
                    const auto& hist = cell_hists[y + by][x + bx];
                    block_feat.insert(block_feat.end(), hist.begin(), hist.end());
                    for (float val : hist) {
                        norm += val * val;
                    }
                }
            }

            // L2 normalization
            norm = static_cast<float>(sqrt(norm + 1e-6));
            for (float& val : block_feat) {
                val /= norm;
            }

            features.insert(features.end(), block_feat.begin(), block_feat.end());
        }
    }
}

void lbp_features(const Mat& src,
                 Mat& dst,
                 int radius,
                 int neighbors) {
    CV_Assert(!src.empty());

    // Convert to grayscale
    Mat gray;
    if (src.channels() == 3) {
        cvtColor(src, gray, COLOR_BGR2GRAY);
    } else {
        gray = src.clone();
    }

    dst = Mat::zeros(gray.size(), CV_8U);

    #pragma omp parallel for
    for (int y = radius; y < gray.rows - radius; y++) {
        for (int x = radius; x < gray.cols - radius; x++) {
            uchar center = gray.at<uchar>(y, x);
            uchar code = 0;

            for (int n = 0; n < neighbors; n++) {
                double theta = 2.0 * PI * n / neighbors;
                int rx = static_cast<int>(x + radius * cos(theta) + 0.5);
                int ry = static_cast<int>(y - radius * sin(theta) + 0.5);

                code |= (gray.at<uchar>(ry, rx) >= center) << n;
            }

            dst.at<uchar>(y, x) = code;
        }
    }
}

void haar_features(const Mat& src,
                  vector<float>& features,
                  Size min_size,
                  Size max_size) {
    CV_Assert(!src.empty());

    // Convert to grayscale
    Mat gray;
    if (src.channels() == 3) {
        cvtColor(src, gray, COLOR_BGR2GRAY);
    } else {
        gray = src.clone();
    }

    // Calculate integral image
    Mat integral;
    compute_integral_image(gray, integral);

    features.clear();

    // Calculate Haar features of different sizes
    for (int h = min_size.height; h <= max_size.height; h += 4) {
        for (int w = min_size.width; w <= max_size.width; w += 4) {
            // Vertical edge features
            for (int y = 0; y <= gray.rows - h; y++) {
                for (int x = 0; x <= gray.cols - w; x++) {
                    int w2 = w / 2;
                    float left = static_cast<float>(integral.at<double>(y + h, x + w2) +
                                                  integral.at<double>(y, x) -
                                                  integral.at<double>(y, x + w2) -
                                                  integral.at<double>(y + h, x));

                    float right = static_cast<float>(integral.at<double>(y + h, x + w) +
                                                   integral.at<double>(y, x + w2) -
                                                   integral.at<double>(y, x + w) -
                                                   integral.at<double>(y + h, x + w2));

                    features.push_back(right - left);
                }
            }

            // Horizontal edge features
            for (int y = 0; y <= gray.rows - h; y++) {
                for (int x = 0; x <= gray.cols - w; x++) {
                    int h2 = h / 2;
                    float top = static_cast<float>(integral.at<double>(y + h2, x + w) +
                                                 integral.at<double>(y, x) -
                                                 integral.at<double>(y, x + w) -
                                                 integral.at<double>(y + h2, x));

                    float bottom = static_cast<float>(integral.at<double>(y + h, x + w) +
                                                    integral.at<double>(y + h2, x) -
                                                    integral.at<double>(y + h2, x + w) -
                                                    integral.at<double>(y + h, x));

                    features.push_back(bottom - top);
                }
            }
        }
    }
}

vector<Mat> create_gabor_filters(int scales,
                               int orientations,
                               Size size) {
    vector<Mat> filters;
    double sigma = 1.0;
    double lambda = 4.0;
    double gamma = 0.5;
    double psi = 0;

    for (int s = 0; s < scales; s++) {
        for (int o = 0; o < orientations; o++) {
            Mat kernel = Mat::zeros(size, CV_32F);
            double theta = o * PI / orientations;
            double sigma_x = sigma;
            double sigma_y = sigma / gamma;

            for (int y = -size.height/2; y <= size.height/2; y++) {
                for (int x = -size.width/2; x <= size.width/2; x++) {
                    double x_theta = x * cos(theta) + y * sin(theta);
                    double y_theta = -x * sin(theta) + y * cos(theta);

                    double gaussian = exp(-0.5 * (x_theta * x_theta / (sigma_x * sigma_x) +
                                                y_theta * y_theta / (sigma_y * sigma_y)));
                    double wave = cos(2 * PI * x_theta / lambda + psi);

                    kernel.at<float>(y + size.height/2, x + size.width/2) =
                        static_cast<float>(gaussian * wave);
                }
            }

            filters.push_back(kernel);
        }

        sigma *= 2;
        lambda *= 2;
    }

    return filters;
}

void gabor_features(const Mat& src,
                   vector<float>& features,
                   int scales,
                   int orientations) {
    CV_Assert(!src.empty());

    // Convert to grayscale
    Mat gray;
    if (src.channels() == 3) {
        cvtColor(src, gray, COLOR_BGR2GRAY);
    } else {
        gray = src.clone();
    }
    gray.convertTo(gray, CV_32F);

    // Create Gabor filter bank
    vector<Mat> filters = create_gabor_filters(scales, orientations);

    features.clear();

    // Apply filters and extract features
    for (const Mat& filter : filters) {
        Mat response;
        filter2D(gray, response, CV_32F, filter);

        // Calculate statistical features of response
        Scalar mean, stddev;
        meanStdDev(response, mean, stddev);

        features.push_back(static_cast<float>(mean[0]));
        features.push_back(static_cast<float>(stddev[0]));
    }
}

void color_histogram(const Mat& src,
                    Mat& hist,
                    const vector<int>& bins) {
    CV_Assert(!src.empty() && src.channels() == 3);

    // Calculate histogram ranges for each channel
    vector<float> ranges[] = {
        vector<float>(bins[0] + 1),
        vector<float>(bins[1] + 1),
        vector<float>(bins[2] + 1)
    };

    for (int i = 0; i < 3; i++) {
        float step = 256.0f / static_cast<float>(bins[i]);
        for (int j = 0; j <= bins[i]; j++) {
            ranges[i][j] = static_cast<float>(j) * step;
        }
    }

    // Split channels
    vector<Mat> channels;
    split(src, channels);

    // Calculate 3D histogram
    int dims[] = {bins[0], bins[1], bins[2]};
    hist = Mat::zeros(3, dims, CV_32F);

    #pragma omp parallel for
    for (int b = 0; b < bins[0]; b++) {
        for (int g = 0; g < bins[1]; g++) {
            for (int r = 0; r < bins[2]; r++) {
                float count = 0.0f;

                for (int y = 0; y < src.rows; y++) {
                    for (int x = 0; x < src.cols; x++) {
                        uchar b_val = channels[0].at<uchar>(y, x);
                        uchar g_val = channels[1].at<uchar>(y, x);
                        uchar r_val = channels[2].at<uchar>(y, x);

                        if (b_val >= ranges[0][b] && b_val < ranges[0][b+1] &&
                            g_val >= ranges[1][g] && g_val < ranges[1][g+1] &&
                            r_val >= ranges[2][r] && r_val < ranges[2][r+1]) {
                            count += 1.0f;
                        }
                    }
                }

                hist.at<float>(b, g, r) = count;
            }
        }
    }

    // Normalize
    normalize(hist, hist, 1, 0, NORM_L1);
}

void compute_integral_image(const Mat& src,
                          Mat& integral) {
    CV_Assert(!src.empty());

    integral.create(src.rows + 1, src.cols + 1, CV_64F);
    integral = Scalar(0);

    // Calculate integral image
    for (int y = 0; y < src.rows; y++) {
        double row_sum = 0;
        for (int x = 0; x < src.cols; x++) {
            row_sum += src.at<uchar>(y, x);
            integral.at<double>(y + 1, x + 1) =
                row_sum + integral.at<double>(y, x + 1);
        }
    }
}

void compute_gradient_histogram(const Mat& magnitude,
                              const Mat& angle,
                              vector<float>& hist,
                              int bin_num) {
    hist.resize(bin_num);
    fill(hist.begin(), hist.end(), 0.0f);

    float bin_size = static_cast<float>(PI) / static_cast<float>(bin_num);

    #pragma omp parallel for
    for (int y = 0; y < magnitude.rows; y++) {
        for (int x = 0; x < magnitude.cols; x++) {
            float mag = magnitude.at<float>(y, x);
            float ang = angle.at<float>(y, x);
            if (ang < 0) ang += static_cast<float>(PI);

            int bin = static_cast<int>(ang / bin_size);
            if (bin >= bin_num) bin = bin_num - 1;

            #pragma omp atomic
            hist[bin] += mag;
        }
    }

    // Normalize
    float sum = 0.0f;
    for (float val : hist) {
        sum += val * val;
    }
    sum = static_cast<float>(sqrt(static_cast<double>(sum) + 1e-6));

    for (float& val : hist) {
        val /= sum;
    }
}

} // namespace ip101

File: python/basic/image_features.py
"""
图像特征相关问题：
51. HOG特征提取 - 使用方向梯度直方图提取特征
52. LBP特征提取 - 使用局部二值模式提取特征
53. Haar特征提取 - 使用Haar-like特征提取特征
54. Gabor特征提取 - 使用Gabor滤波器提取特征
55. 颜色直方图 - 计算图像的颜色直方图特征
"""

import numpy as np
from scipy import ndimage
import cv2
import matplotlib.pyplot as plt
import math

def hog_features(img_path, cell_size=8, block_size=2, bins=9):
    """
    问题51：HOG特征提取
    使用方向梯度直方图提取特征

    参数:
        img_path: 输入图像路径
        cell_size: 单元格大小，默认为8
        block_size: 块大小（以cell为单位），默认为2
        bins: 方向直方图的bin数量，默认为9

    返回:
        HOG特征可视化结果
    """
    # 读取图像
    img = cv2.imread(img_path, cv2.IMREAD_GRAYSCALE)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    # 计算梯度
    gx = cv2.Sobel(img, cv2.CV_32F, 1, 0)
    gy = cv2.Sobel(img, cv2.CV_32F, 0, 1)
    magnitude = np.sqrt(gx**2 + gy**2)
    orientation = np.arctan2(gy, gx) * 180 / np.pi % 180

    # 计算cell直方图
    h, w = img.shape
    h_cells = h // cell_size
    w_cells = w // cell_size
    histogram = np.zeros((h_cells, w_cells, bins))

    for i in range(h_cells):
        for j in range(w_cells):
            cell_mag = magnitude[i*cell_size:(i+1)*cell_size,
                               j*cell_size:(j+1)*cell_size]
            cell_ori = orientation[i*cell_size:(i+1)*cell_size,
                                 j*cell_size:(j+1)*cell_size]

            # 计算直方图
            for m in range(cell_size):
                for n in range(cell_size):
                    ori = cell_ori[m, n]
                    mag = cell_mag[m, n]
                    bin_index = int(ori / 180 * bins)
                    histogram[i, j, bin_index] += mag

    # 可视化HOG特征
    result = np.zeros_like(img)
    for i in range(h_cells):
        for j in range(w_cells):
            for k in range(bins):
                angle = k * 180 / bins
                rho = histogram[i, j, k]
                x = int((j + 0.5) * cell_size)
                y = int((i + 0.5) * cell_size)
                dx = int(rho * np.cos(angle * np.pi / 180))
                dy = int(rho * np.sin(angle * np.pi / 180))
                cv2.line(result, (x, y), (x + dx, y + dy), 255, 1)

    # 转换为彩色图像
    result = cv2.cvtColor(result, cv2.COLOR_GRAY2BGR)

    return result

def lbp_features(img_path):
    """
    问题52：LBP特征提取
    使用局部二值模式提取特征

    参数:
        img_path: 输入图像路径

    返回:
        LBP特征图像
    """
    # 读取图像
    img = cv2.imread(img_path, cv2.IMREAD_GRAYSCALE)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    h, w = img.shape
    result = np.zeros((h-2, w-2), dtype=np.uint8)

    # 计算LBP特征
    for i in range(1, h-1):
        for j in range(1, w-1):
            center = img[i, j]
            code = 0
            code |= (img[i-1, j-1] > center) << 7
            code |= (img[i-1, j] > center) << 6
            code |= (img[i-1, j+1] > center) << 5
            code |= (img[i, j+1] > center) << 4
            code |= (img[i+1, j+1] > center) << 3
            code |= (img[i+1, j] > center) << 2
            code |= (img[i+1, j-1] > center) << 1
            code |= (img[i, j-1] > center) << 0
            result[i-1, j-1] = code

    # 转换为彩色图像
    result = cv2.cvtColor(result, cv2.COLOR_GRAY2BGR)

    return result

def haar_features(img_path):
    """
    问题53：Haar特征提取
    使用Haar-like特征提取特征

    参数:
        img_path: 输入图像路径

    返回:
        Haar特征可视化结果
    """
    # 读取图像
    img = cv2.imread(img_path, cv2.IMREAD_GRAYSCALE)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    # 计算积分图
    integral = cv2.integral(img)

    # 定义Haar特征类型
    def compute_haar_feature(x, y, w, h, feature_type):
        if feature_type == 'edge_horizontal':
            # 水平边缘特征
            top = integral[y+h//2, x+w] - integral[y+h//2, x] - integral[y, x+w] + integral[y, x]
            bottom = integral[y+h, x+w] - integral[y+h, x] - integral[y+h//2, x+w] + integral[y+h//2, x]
            return abs(top - bottom)
        elif feature_type == 'edge_vertical':
            # 垂直边缘特征
            left = integral[y+h, x+w//2] - integral[y+h, x] - integral[y, x+w//2] + integral[y, x]
            right = integral[y+h, x+w] - integral[y+h, x+w//2] - integral[y, x+w] + integral[y, x+w//2]
            return abs(left - right)
        return 0

    # 计算特征响应
    result = np.zeros_like(img)
    window_size = 24
    step = 4

    for y in range(0, img.shape[0]-window_size, step):
        for x in range(0, img.shape[1]-window_size, step):
            # 计算水平和垂直边缘特征
            h_response = compute_haar_feature(x, y, window_size, window_size, 'edge_horizontal')
            v_response = compute_haar_feature(x, y, window_size, window_size, 'edge_vertical')
            response = (h_response + v_response) / 2
            result[y:y+window_size, x:x+window_size] = response

    # 归一化结果
    result = cv2.normalize(result, None, 0, 255, cv2.NORM_MINMAX).astype(np.uint8)

    # 转换为彩色图像
    result = cv2.cvtColor(result, cv2.COLOR_GRAY2BGR)

    return result

def gabor_features(img_path, ksize=31, sigma=4.0, theta=0, lambd=10.0, gamma=0.5):
    """
    问题54：Gabor特征提取
    使用Gabor滤波器提取特征

    参数:
        img_path: 输入图像路径
        ksize: 核大小
        sigma: 标准差
        theta: 方向
        lambd: 波长
        gamma: 空间纵横比

    返回:
        Gabor特征图像
    """
    # 读取图像
    img = cv2.imread(img_path, cv2.IMREAD_GRAYSCALE)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    # 创建Gabor滤波器组
    filters = []
    num_theta = 8  # 方向数量
    for theta in np.arange(0, np.pi, np.pi/num_theta):
        kernel = cv2.getGaborKernel((ksize, ksize), sigma, theta, lambd, gamma, 0, ktype=cv2.CV_32F)
        filters.append(kernel)

    # 应用滤波器
    responses = []
    for kernel in filters:
        response = cv2.filter2D(img, cv2.CV_32F, kernel)
        responses.append(response)

    # 合并响应
    result = np.max(responses, axis=0)
    result = cv2.normalize(result, None, 0, 255, cv2.NORM_MINMAX).astype(np.uint8)

    # 转换为彩色图像
    result = cv2.cvtColor(result, cv2.COLOR_GRAY2BGR)

    return result

def color_histogram(img_path, bins=32):
    """
    问题55：颜色直方图
    计算图像的颜色直方图特征

    参数:
        img_path: 输入图像路径
        bins: 直方图bin数量，默认为32

    返回:
        颜色直方图可视化结果
    """
    # 读取图像
    img = cv2.imread(img_path)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    # 转换到HSV色彩空间
    hsv = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)

    # 计算直方图
    hist_h = cv2.calcHist([hsv], [0], None, [bins], [0, 180])
    hist_s = cv2.calcHist([hsv], [1], None, [bins], [0, 256])
    hist_v = cv2.calcHist([hsv], [2], None, [bins], [0, 256])

    # 归一化直方图
    cv2.normalize(hist_h, hist_h, 0, 255, cv2.NORM_MINMAX)
    cv2.normalize(hist_s, hist_s, 0, 255, cv2.NORM_MINMAX)
    cv2.normalize(hist_v, hist_v, 0, 255, cv2.NORM_MINMAX)

    # 创建直方图可视化
    height = 400
    width = bins * 3
    result = np.zeros((height, width, 3), dtype=np.uint8)

    # 绘制直方图
    for i in range(bins):
        # H通道 - 红色
        cv2.line(result, (i*3, height), (i*3, height-int(hist_h[i])), (0,0,255), 1)
        # S通道 - 绿色
        cv2.line(result, (i*3+1, height), (i*3+1, height-int(hist_s[i])), (0,255,0), 1)
        # V通道 - 蓝色
        cv2.line(result, (i*3+2, height), (i*3+2, height-int(hist_v[i])), (255,0,0), 1)

    return result

def compute_hog_manual(image, cell_size=8, block_size=2, bins=9):
    """
    手动实现HOG特征提取

    参数:
        image: 输入图像(灰度图)
        cell_size: 每个cell的大小
        block_size: 每个block包含的cell数量
        bins: 方向梯度直方图的bin数量

    返回:
        hog_features: HOG特征向量
    """
    # 1. 计算图像梯度
    dx = ndimage.sobel(image, axis=1)
    dy = ndimage.sobel(image, axis=0)

    # 2. 计算梯度幅值和方向
    magnitude = np.sqrt(dx**2 + dy**2)
    orientation = np.arctan2(dy, dx) * 180 / np.pi % 180

    # 3. 计算cell的梯度直方图
    cell_rows = image.shape[0] // cell_size
    cell_cols = image.shape[1] // cell_size
    histogram = np.zeros((cell_rows, cell_cols, bins))

    for i in range(cell_rows):
        for j in range(cell_cols):
            # 获取当前cell的梯度和方向
            cell_mag = magnitude[i*cell_size:(i+1)*cell_size,
                               j*cell_size:(j+1)*cell_size]
            cell_ori = orientation[i*cell_size:(i+1)*cell_size,
                                 j*cell_size:(j+1)*cell_size]

            # 计算投票权重
            for m in range(cell_size):
                for n in range(cell_size):
                    ori = cell_ori[m, n]
                    mag = cell_mag[m, n]

                    # 双线性插值投票
                    bin_index = int(ori / 180 * bins)
                    bin_index_next = (bin_index + 1) % bins
                    weight_next = (ori - bin_index * 180 / bins) / (180 / bins)
                    weight = 1 - weight_next

                    histogram[i, j, bin_index] += mag * weight
                    histogram[i, j, bin_index_next] += mag * weight_next

    # 4. Block归一化
    blocks_rows = cell_rows - block_size + 1
    blocks_cols = cell_cols - block_size + 1
    normalized_blocks = np.zeros((blocks_rows, blocks_cols,
                                block_size * block_size * bins))

    for i in range(blocks_rows):
        for j in range(blocks_cols):
            block = histogram[i:i+block_size, j:j+block_size, :].ravel()
            normalized_blocks[i, j, :] = block / np.sqrt(np.sum(block**2) + 1e-6)

    return normalized_blocks.ravel()

def compute_lbp_manual(image, points=8, radius=1):
    """
    手动实现LBP特征提取

    参数:
        image: 输入图像(灰度图)
        points: 采样点数量
        radius: 采样半径

    返回:
        lbp_image: LBP特征图
    """
    rows = image.shape[0]
    cols = image.shape[1]
    lbp_image = np.zeros_like(image)

    for i in range(radius, rows-radius):
        for j in range(radius, cols-radius):
            center = image[i, j]
            binary = ''

            # 计算采样点的坐标和值
            for p in range(points):
                angle = 2 * np.pi * p / points
                x = i + radius * np.cos(angle)
                y = j - radius * np.sin(angle)

                # 双线性插值
                x1, y1 = int(x), int(y)
                x2, y2 = math.ceil(x), math.ceil(y)
                fx, fy = x - x1, y - y1

                value = (1-fx)*(1-fy)*image[x1,y1] + \
                        fx*(1-fy)*image[x2,y1] + \
                        (1-fx)*fy*image[x1,y2] + \
                        fx*fy*image[x2,y2]

                binary += '1' if value >= center else '0'

            lbp_image[i,j] = int(binary, 2)

    return lbp_image

def compute_haar_manual(image, feature_type='edge'):
    """手动实现Haar特征提取

    参数:
        image: 输入的灰度图像
        feature_type: Haar特征类型，可选值为'edge'(边缘),'line'(线),'center'(中心)

    返回:
        feature_map: Haar特征图
    """
    if len(image.shape) > 2:
        image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

    height, width = image.shape
    feature_map = np.zeros((height, width))

    # 根据特征类型选择不同的计算方法
    if feature_type == 'edge':
        # 计算水平边缘特征
        for y in range(height-1):
            for x in range(width):
                feature_map[y,x] = float(image[y+1,x]) - float(image[y,x])

    elif feature_type == 'line':
        # 计算垂直线特征
        for y in range(height):
            for x in range(width-2):
                feature_map[y,x] = float(image[y,x]) + float(image[y,x+2]) - 2*float(image[y,x+1])

    elif feature_type == 'center':
        # 计算中心特征
        for y in range(1, height-1):
            for x in range(1, width-1):
                center = float(image[y,x])
                surroundings = (float(image[y-1,x]) + float(image[y+1,x]) +
                              float(image[y,x-1]) + float(image[y,x+1])) / 4
                feature_map[y,x] = center - surroundings

    return feature_map

def compute_gabor_manual(image, ksize=31, sigma=4.0, theta=0, lambda_=10.0, gamma=0.5, psi=0):
    """手动实现Gabor滤波器特征提取

    参数:
        image: 输入的灰度图像
        ksize: 滤波器核的大小
        sigma: 高斯包络的标准差
        theta: 滤波器的方向(弧度)
        lambda_: 正弦波的波长
        gamma: 空间纵横比
        psi: 相位偏移

    返回:
        feature_map: Gabor特征图
    """
    if len(image.shape) > 2:
        image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

    # 确保ksize是奇数
    ksize = int(ksize)
    if ksize % 2 == 0:
        ksize += 1

    # 创建网格坐标
    k = (ksize-1) // 2
    y, x = np.mgrid[-k:k+1, -k:k+1]

    # 旋转坐标
    x_theta = x * np.cos(theta) + y * np.sin(theta)
    y_theta = -x * np.sin(theta) + y * np.cos(theta)

    # 计算Gabor滤波器核
    gb = np.exp(-.5 * (x_theta**2 + gamma**2 * y_theta**2) / sigma**2)
    gb *= np.cos(2*np.pi*x_theta/lambda_ + psi)
    gb /= 2*np.pi*sigma**2

    # 应用滤波器
    feature_map = ndimage.convolve(image.astype(float), gb)

    return feature_map

def compute_color_histogram(image, bins=32):
    """计算颜色直方图特征

    参数:
        image: 输入的RGB图像
        bins: 每个颜色通道的直方图柱数

    返回:
        features: 归一化后的颜色直方图特征向量
    """
    if len(image.shape) < 3:
        raise ValueError("输入图像必须是RGB图像")

    features = []

    # 对每个颜色通道计算直方图
    for i in range(3):
        hist = cv2.calcHist([image], [i], None, [bins], [0, 256])
        hist = cv2.normalize(hist, hist).flatten()
        features.extend(hist)

    return np.array(features)

def compute_harris_manual(image, k=0.04, window_size=3, threshold=0.01):
    """手动实现Harris角点检测

    参数:
        image: 输入的灰度图像
        k: Harris响应函数参数，默认0.04
        window_size: 局部窗口大小，默认3
        threshold: 角点检测阈值，默认0.01

    返回:
        corners: 角点检测结果图像
    """
    if len(image.shape) > 2:
        image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

    # 计算x和y方向的梯度
    dx = ndimage.sobel(image, axis=1)
    dy = ndimage.sobel(image, axis=0)

    # 计算梯度乘积
    Ixx = dx * dx
    Ixy = dx * dy
    Iyy = dy * dy

    # 使用高斯窗口进行平滑
    window = np.ones((window_size, window_size)) / (window_size * window_size)
    Sxx = ndimage.convolve(Ixx, window)
    Sxy = ndimage.convolve(Ixy, window)
    Syy = ndimage.convolve(Iyy, window)

    # 计算Harris响应
    det = Sxx * Syy - Sxy * Sxy
    trace = Sxx + Syy
    harris_response = det - k * (trace * trace)

    # 阈值处理
    corners = np.zeros_like(image)
    corners[harris_response > threshold * harris_response.max()] = 255

    return corners

def main(problem_id=51):
    """
    主函数，通过problem_id选择要运行的问题

    参数:
        problem_id: 问题编号，默认为51
    """
    input_path = "../images/imori.jpg"

    if problem_id == 51:
        result = hog_features(input_path)
        output_path = "../images/answer_51.jpg"
        title = "HOG特征"
    elif problem_id == 52:
        result = lbp_features(input_path)
        output_path = "../images/answer_52.jpg"
        title = "LBP特征"
    elif problem_id == 53:
        result = haar_features(input_path)
        output_path = "../images/answer_53.jpg"
        title = "Haar特征"
    elif problem_id == 54:
        result = gabor_features(input_path)
        output_path = "../images/answer_54.jpg"
        title = "Gabor特征"
    elif problem_id == 55:
        result = color_histogram(input_path)
        output_path = "../images/answer_55.jpg"
        title = "颜色直方图"
    else:
        print(f"问题 {problem_id} 不存在")
        return

    # 保存结果
    cv2.imwrite(output_path, result)

    # 显示结果
    cv2.imshow(title, result)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

if __name__ == "__main__":
    import sys
    problem_id = int(sys.argv[1]) if len(sys.argv) > 1 else 51
    main(problem_id)